### PR TITLE
adjust TilemapTextureArray mipmap levels to remove most aliasing

### PR DESCRIPTION
**Pull request by petchema**
I did perceptual tests on modded terrain textures, then some analysis, using DREAM RC4 and no post processing mod on my screen (1080p).

First I checked that a mipmap level of 8.0 looked smooth in the distance: that works.
Then I checked how far the transition between level 7.0 and 8.0 could be pushed away before aliasing appears. In my tests, 50.0 was about right.
Then I checked how far the transition between level 0.0 and 7.0 could be pushed away before aliasing appears. In my tests, 4.5 was close to the max.

But what's the scale between the two? I assume the goal is to keep the angular diameter of pixels about constant. Pixel widths varies according to 1/d, but their height decreases faster because the surface also appears more and more tilted (from camera point of view).
I spare you the details, but height seems to vary in 1/d².

Mipmap scale is logarithmic (vs pixel size), so you end up with level = const - 2*log2(d)

In other words, you should gain one mipmap level each time the distance increases by a sqrt(2) factor.
And it matches almost perfectly with the two bounds found perceptually!

However I don't know how my perceptual tests are portable to other systems...

Forums: https://forums.dfworkshop.net/viewtopic.php?f=27&t=1168&p=25573#p25573

### CODE REVIEW

**Comment by petchema** [comment]
It gives terrible blurriness with unmodded terrain textures. I guess it must be adjusted depending on textures resolution?
Either that, or separate shaders should be used for modded and unmodded textures...

**Comment by Nystul-the-Magician** [comment]
we might be able to modify the shader a bit and use the texture size property there - see https://docs.unity3d.com/Manual/SL-PropertiesInPrograms.html

**Comment by petchema** [comment]
Any way to optimize the computation of scale (or even 1.0f / scale)? It should never change for a given texture...

**Comment by petchema** [comment]
I found another issue...
The terrain may look good on foot or on horse, but if you look at it from high above, the hypothesis that you're looking tangentially doesn't hold, and it looks terribly blurry again :(
mipmap-dream-nopostprocessing

**Comment by petchema** [comment]
Fixed!
heightfix-dream-nopostprocessing

**Comment by Interkarma** [comment]
This looks great! Looking forward to playing with it. Well done. :)

**Comment by petchema** [comment]
It still makes some approximations about the terrain (mainly, that the texture is horizontal) so in some cases, like if you're looking upward at a hill you're climbing, or you're simply looking at the front of a bump, the selected mipmap level is too high.
A surface shader that computes mipmap level from ddx and ddy would give better results, but I don't quite understand the math :(

**Comment by Interkarma** [comment]
My original tilemap shader in 2014 used ddx/ddy to manually calculate mipmap level. Not sure if that will be helpful as a reference or not. Here's the link to the article from way, way back.

https://www.dfworkshop.net/improved-terrain-tiling/

**Comment by Interkarma** [comment]
Very nice improvement, thank you! This makes the new tilemap terrain look every bit as nice in the air as my old terrain shader. Awesome! :)

**Comment by Nystul-the-Magician** [comment]
> 
> 
> My original tilemap shader in 2014 used ddx/ddy to manually calculate mipmap level. Not sure if that will be helpful as a reference or not. Here's the link to the article from way, way back.
> 
> https://www.dfworkshop.net/improved-terrain-tiling/

I am not 100% sure if I remember correctly, but I think back then when I wrote the texture array tilemap shader I pretty much started from your shader with the ddx/ddy computations but as far as I remember I had problems with seams on the edges of tiles so I decided against the ddx/ddy.
It is likely that I did something wrong or it was a unity bug back then, so trying it out again would be a good thing to do!
